core/state/statedb.go

```diff
@@ -347,90 +381,106 @@ func (s *StateDB) Database() Database {
 
 // StorageTrie returns the storage trie of an account.
 // The return value is a copy and is nil for non-existent accounts.
-func (s *StateDB) StorageTrie(addr common.Address) Trie {
+func (s *StateDB) StorageTrie(addr common.Address) (Trie, error) {
+	if !addr.IsInChainScope() {
+		return nil, ErrInvalidScope
+	}
 	stateObject := s.getStateObject(addr)
 	if stateObject == nil {
-		return nil
+		return nil, nil
 	}
 	cpy := stateObject.deepCopy(s)
 	cpy.updateTrie(s.db)
-	return cpy.getTrie(s.db)
+	return cpy.getTrie(s.db), nil
 }
 
-func (s *StateDB) HasSuicided(addr common.Address) bool {
+func (s *StateDB) HasSuicided(addr common.Address) (bool, error) {
+	if !addr.IsInChainScope() {
+		return false, ErrInvalidScope
+	}
 	stateObject := s.getStateObject(addr)
 	if stateObject != nil {
-		return stateObject.suicided
+		return stateObject.suicided, nil
 	}
-	return false
+	return false, nil
 }
 
 /*
  * SETTERS
  */
 
 // AddBalance adds amount to the account associated with addr.
-func (s *StateDB) AddBalance(addr common.Address, amount *big.Int) {
-	stateObject := s.GetOrNewStateObject(addr)
-	if stateObject != nil {
+func (s *StateDB) AddBalance(addr common.Address, amount *big.Int) error {
```

Comment:
We should make addr common.Address an interface so that it can type assert on internal and external type so that we can always guarantee that we use Addresses properly and assert if they are out of context.

Reply:
https://github.com/dominant-strategies/go-quai/issues/519